File: node/koa_graphql/graphql/schema.js
import Post from '../models/post';

import PostType from './types/postType';


import {
  GraphQLBoolean,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
} from 'graphql';

let Schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'RootQueryType',
    fields: () => ({
      post: {
        type: PostType,
        // args will automatically be mapped to `where`
        args: {
          id: {
            description: 'id of the post',
            type: new GraphQLNonNull(GraphQLInt)
          }
        },
        resolve: (obj, { id, ...args }) => Post.forge({ id: id }).fetch({withRelated: ['comments', 'user', 'comments.user'], debug: true}).then((post) => post.toJSON())
      },
      allPosts: {
        // The resolver will use `findOne` or `findAll` depending on whether the field it's used in is a `GraphQLList` or not.
        type: new GraphQLList(PostType),
        args: {
          // An arg with the key limit will automatically be converted to a limit on the target
          limit: {
            type: GraphQLInt
          },
          // An arg with the key order will automatically be converted to a order on the target
          order: {
            type: GraphQLString
          }
        },
        resolve: (obj, { id, ...args }) => Post.forge().fetchAll({withRelated: ['comments', 'user', 'comments.user'], debug: true}).then((posts) => posts.toJSON())
      }
    }),
  })
});


module.exports = Schema;

File: elixir/phoenix_graphql/web/api/schema.ex
defmodule PhoenixGraphql.Api.Schema do
	use Absinthe.Schema
	import_types PhoenixGraphql.Api.Schema.Types.Post
	import_types PhoenixGraphql.Api.Schema.Types.User
	import_types PhoenixGraphql.Api.Schema.Types.Comment
	alias PhoenixGraphql.Api.Resolver


	query do
	  field :posts, list_of(:post) do
	    resolve fn
	    	(_, _) -> Resolver.all(
	    		%{
	    			model: PhoenixGraphql.Post,
	    			preloaders: [:user, :comments,  [comments: :user]]
	    		}
	    	)
	    end
	  end

    field :post, :post do
      arg :id, non_null(:id)
      resolve fn
        %{id: id}, _ -> Resolver.find(
          %{
            id: id,
            model: PhoenixGraphql.Post,
            preloaders: [:user, :comments, [comments: :user]]
          }
        )
      end
    end

	  field :users, list_of(:user) do
	    resolve fn
	    	(_, _) -> Resolver.all(
	    		%{
	    			model: PhoenixGraphql.User,
	    			preloaders: []
	    		}
	    	)
	    end
	  end

	  field :user, :user do
	    arg :id, non_null(:id)
	    resolve fn
        %{id: id}, _ -> {:ok, PhoenixGraphql.Repo.get(PhoenixGraphql.User, id) |> PhoenixGraphql.Repo.preload([:comments, :posts])}
      end
	  end
	end
end
